docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,438 @@
+    `PrepareProposal` contain their corresponding vote extensions. In other words, we render vote
+    extensions optional.
+    There are strong implications stemming from such a relaxation of the original specification.
+
+    - As a vote extension is signed *separately* from the vote it is extending, an intermediate node
+      can now remove (i.e., censor) vote extensions from precommit messages at will.
+    - Further, there is no point anymore in the spec requiring the Application to accept a vote extension
+      passed via `VerifyVoteExtension` to consider a precommit message valid in its entirety. Remember
+      this behavior of `VerifyVoteExtension` is adding a constraint to Tendermint's conditions for
+      liveness.
+      In this situation, it is better and simpler to just drop the vote extension rejected by the
+      Application via `VerifyVoteExtension`, but still consider the vote itself valid as long as its
+      signature verifies.
+
+- **1.** *Include vote extensions in the blockchain*.
+
+    Another obvious solution, which has somehow been considered in the past, is to include the vote
+    extensions and their signatures in the blockchain.
+    The blockchain would thus include the extended commit, rather than a regular commit, as the structure
+    to be canonicalized in the next block.
+    With this solution, the current mechanisms implemented both in the blocksync and consensus reactors
+    would still be correct, as all the information a node needs to catch up, and to start proposing when
+    it considers it is caught up, can now be recovered from past blocks.
+
+    This solution has two main drawbacks.
+
+    - It changes the blockchain format, in particular, the way light blocks are verified; thus breaking
+      backwards compatiblity in features such as light clients and IBC (the latter relying on the former).
+    - The extra information (i.e., the vote extensions) that is now kept in the blockchain is not really
+      needed *at every height* for a late node to catch up. This information is only needed to be able
+      to *propose* at the height the validator considers itself as caught-up. If a validator is indeed
+      late for height *h*, it is useless (although correct) for it to call `PrepareProposal`, or
+      `ExtendVote`, since the block is already decided.
+
+- **2.** *Skip* propose *step in Tendermint algorithm*.
+
+    This solution consists in modifying the Tendermint algorithm to skip the *send proposal* step in
+    heights where the node does not have the required vote extensions to populate the call to
+    `PrepareProposal`. The assumption is that the validator is late and, therefore, up-to-date
+    validators have already proposed (and decided) for that height.
+    A small variation of this solution is, rather than skipping the *send proposal* step, the validator
+    sends a special *empty* or *bottom* proposal to signal other nodes that it is not ready to propose
```

Comment:
Yes, you can call it *bottom* or *nil*. And yes, the idea is you would always prevote `nil` for it.
However, I didn't want to expand on this here as, I think, this is not the good solution... until you convince me otherwise :-)
Let me know if you still think I should expand on this here.

Reply:
I think that this solution is somewhat unrelated to the problem of Vote Extensions. It's a very cool idea in general, since it effectively should work to speed up consensus in cases where a proposer does not want to commit, but it can / should be considered separately.